File: python/src/io.cpp
// Copyright (C) 2017-2019 Chris N. Richardson Garth N. Wells
//
// This file is part of DOLFIN (https://www.fenicsproject.org)
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include "casters.h"
#include <dolfin/function/Function.h>
#include <dolfin/function/FunctionSpace.h>
#include <dolfin/geometry/Point.h>
#include <dolfin/io/HDF5File.h>
#include <dolfin/io/XDMFFile.h>
#include <dolfin/la/PETScVector.h>
#include <dolfin/mesh/Mesh.h>
#include <dolfin/mesh/MeshFunction.h>
#include <dolfin/mesh/MeshValueCollection.h>
#include <memory>
#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <string>
#include <vector>

namespace py = pybind11;

namespace dolfin_wrappers
{

void io(py::module& m)
{
  // dolfin::io::HDF5File
  py::class_<dolfin::io::HDF5File, std::shared_ptr<dolfin::io::HDF5File>>(
      m, "HDF5File", py::dynamic_attr())
      .def(py::init([](const MPICommWrapper comm, const std::string filename,
                       const std::string file_mode) {
             return std::make_unique<dolfin::io::HDF5File>(comm.get(), filename,
                                                           file_mode);
           }),
           py::arg("comm"), py::arg("filename"), py::arg("file_mode"))
      .def("close", &dolfin::io::HDF5File::close)
      .def("flush", &dolfin::io::HDF5File::flush)
      // read
      .def("read_mesh",
           [](dolfin::io::HDF5File& self, const MPICommWrapper comm,
              const std::string data_path, bool use_partition_from_file,
              const dolfin::mesh::GhostMode ghost_mode) {
             return self.read_mesh(comm.get(), data_path,
                                   use_partition_from_file, ghost_mode);
           })
      .def("read_vector",
           [](dolfin::io::HDF5File& self, const MPICommWrapper comm,
              const std::string data_path, bool use_partition_from_file) {
             auto x = self.read_vector(comm.get(), data_path,
                                       use_partition_from_file);
             Vec _x = x.vec();
             PetscObjectReference((PetscObject)_x);
             return _x;
           },
           py::return_value_policy::take_ownership)
      .def("read_mf_bool", &dolfin::io::HDF5File::read_mf_bool, py::arg("mesh"),
           py::arg("name"))
      .def("read_mf_int", &dolfin::io::HDF5File::read_mf_int, py::arg("mesh"),
           py::arg("name"))
      .def("read_mf_size_t", &dolfin::io::HDF5File::read_mf_size_t,
           py::arg("mesh"), py::arg("name"))
      .def("read_mf_double", &dolfin::io::HDF5File::read_mf_double,
           py::arg("mesh"), py::arg("name"))
      //
      .def("read_mvc_bool", &dolfin::io::HDF5File::read_mvc_bool,
           py::arg("mesh"), py::arg("name"))
      .def("read_mvc_size_t", &dolfin::io::HDF5File::read_mvc_size_t,
           py::arg("mesh"), py::arg("name"))
      .def("read_mvc_double", &dolfin::io::HDF5File::read_mvc_double,
           py::arg("mesh"), py::arg("name"))
      .def("read",
           py::overload_cast<
               std::shared_ptr<const dolfin::function::FunctionSpace>,
               const std::string>(&dolfin::io::HDF5File::read, py::const_),
           py::arg("V"), py::arg("name"))
      // write
      .def("write", (void (dolfin::io::HDF5File::*)(const dolfin::mesh::Mesh&,
                                                    std::string))
                        & dolfin::io::HDF5File::write)
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshValueCollection<bool>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("mvc"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshValueCollection<std::size_t>&,
               std::string))
               & dolfin::io::HDF5File::write,
           py::arg("mvc"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshValueCollection<double>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("mvc"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshFunction<bool>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("meshfunction"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshFunction<std::size_t>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("meshfunction"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshFunction<int>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("meshfunction"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(
               const dolfin::mesh::MeshFunction<double>&, std::string))
               & dolfin::io::HDF5File::write,
           py::arg("meshfunction"), py::arg("name"))
      .def("write",
           [](dolfin::io::HDF5File& self, Vec x, std::string s) {
             dolfin::la::PETScVector _x(x);
             self.write(_x, s);
           },
           py::arg("vector"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(const dolfin::function::Function&,
                                           std::string))
               & dolfin::io::HDF5File::write,
           py::arg("u"), py::arg("name"))
      .def("write",
           (void (dolfin::io::HDF5File::*)(const dolfin::function::Function&,
                                           std::string, double))
               & dolfin::io::HDF5File::write,
           py::arg("u"), py::arg("name"), py::arg("t"))
      .def("set_mpi_atomicity", &dolfin::io::HDF5File::set_mpi_atomicity)
      .def("get_mpi_atomicity", &dolfin::io::HDF5File::get_mpi_atomicity)
      .def_readwrite("chunking", &dolfin::io::HDF5File::chunking)
      // others
      .def("has_dataset", &dolfin::io::HDF5File::has_dataset);

  // dolfin::io::XDMFFile
  py::class_<dolfin::io::XDMFFile, std::shared_ptr<dolfin::io::XDMFFile>>
      xdmf_file(m, "XDMFFile");

  xdmf_file
      .def(py::init([](const MPICommWrapper comm, std::string filename,
                       dolfin::io::XDMFFile::Encoding encoding) {
             return std::make_unique<dolfin::io::XDMFFile>(comm.get(), filename,
                                                           encoding);
           }),
           py::arg("comm"), py::arg("filename"), py::arg("encoding"))
      .def("close", &dolfin::io::XDMFFile::close)
      .def_readwrite("functions_share_mesh",
                     &dolfin::io::XDMFFile::functions_share_mesh)
      .def_readwrite("flush_output", &dolfin::io::XDMFFile::flush_output)
      .def_readwrite("rewrite_function_mesh",
                     &dolfin::io::XDMFFile::rewrite_function_mesh);

  // dolfin::io::XDMFFile::Encoding enums
  py::enum_<dolfin::io::XDMFFile::Encoding>(xdmf_file, "Encoding")
      .value("HDF5", dolfin::io::XDMFFile::Encoding::HDF5)
      .value("ASCII", dolfin::io::XDMFFile::Encoding::ASCII);

  // dolfin::io::XDMFFile::write
  xdmf_file
      // Function
      .def("write",
           py::overload_cast<const dolfin::function::Function&>(
               &dolfin::io::XDMFFile::write),
           py::arg("u"))
      .def("write",
           py::overload_cast<const dolfin::function::Function&, double>(
               &dolfin::io::XDMFFile::write),
           py::arg("u"), py::arg("t"))
      // Mesh
      .def("write",
           py::overload_cast<const dolfin::mesh::Mesh&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mesh"))
      // MeshFunction
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshFunction<bool>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mf"))
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshFunction<std::size_t>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mf"))
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshFunction<int>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mf"))
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshFunction<double>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mf"))
      // MeshValueCollection
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshValueCollection<bool>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mvc"))
      .def("write",
           py::overload_cast<
               const dolfin::mesh::MeshValueCollection<std::size_t>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mvc"))
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshValueCollection<int>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mvc"))
      .def("write",
           py::overload_cast<const dolfin::mesh::MeshValueCollection<double>&>(
               &dolfin::io::XDMFFile::write),
           py::arg("mvc"))
      // Points
      .def("write",
           [](dolfin::io::XDMFFile& instance, py::list points) {
             auto _points = points.cast<std::vector<dolfin::geometry::Point>>();
             instance.write(_points);
           },
           py::arg("points"))
      // Points with values
      .def("write",
           [](dolfin::io::XDMFFile& instance, py::list points,
              std::vector<double>& values) {
             auto _points = points.cast<std::vector<dolfin::geometry::Point>>();
             instance.write(_points, values);
           },
           py::arg("points"), py::arg("values"))
      // Check points
      .def("write_checkpoint",
           [](dolfin::io::XDMFFile& instance,
              const dolfin::function::Function& u, std::string function_name,
              double time_step) {
             instance.write_checkpoint(u, function_name, time_step);
           },
           py::arg("u"), py::arg("function_name"), py::arg("time_step") = 0.0);

  // XDFMFile::read
  xdmf_file
      // Mesh
      .def("read_mesh",
           [](dolfin::io::XDMFFile& self, const MPICommWrapper comm,
              const dolfin::mesh::GhostMode ghost_mode) {
             return self.read_mesh(comm.get(), ghost_mode);
           })
      // MeshFunction
      .def("read_mf_bool", &dolfin::io::XDMFFile::read_mf_bool, py::arg("mesh"),
           py::arg("name") = "")
      .def("read_mf_int", &dolfin::io::XDMFFile::read_mf_int, py::arg("mesh"),
           py::arg("name") = "")
      .def("read_mf_size_t", &dolfin::io::XDMFFile::read_mf_size_t,
           py::arg("mesh"), py::arg("name") = "")
      .def("read_mf_double", &dolfin::io::XDMFFile::read_mf_double,
           py::arg("mesh"), py::arg("name") = "")
      // MeshValueCollection
      .def("read_mvc_bool", &dolfin::io::XDMFFile::read_mvc_bool,
           py::arg("mesh"), py::arg("name") = "")
      .def("read_mvc_int", &dolfin::io::XDMFFile::read_mvc_int, py::arg("mesh"),
           py::arg("name") = "")
      .def("read_mvc_size_t", &dolfin::io::XDMFFile::read_mvc_size_t,
           py::arg("mesh"), py::arg("name") = "")
      .def("read_mvc_double", &dolfin::io::XDMFFile::read_mvc_double,
           py::arg("mesh"), py::arg("name") = "")
      // Checkpointing
      .def("read_checkpoint", &dolfin::io::XDMFFile::read_checkpoint,
           py::arg("V"), py::arg("name"), py::arg("counter") = -1);
}
} // namespace dolfin_wrappers

File: python/dolfin/io.py
# Copyright (C) 2017-2018 Chris N. Richardson, Garth N. Wells and Michal Habera
#
# This file is part of DOLFIN (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    LGPL-3.0-or-later
"""IO module for input data, post-processing and checkpointing"""

import typing

import numpy

from dolfin import cpp, fem, function


__all__ = ["HDF5File", "XDMFFile"]


class HDF5File:
    """Interface to HDF5 files"""

    def __init__(self, mpi_comm, filename: str, mode: str):
        """Open HDF5 file

        Parameters
        ----------
        mpi_comm
            The MPI communicator
        filename
            Name of the file
        mode
            File opening mode, which can be write (w), read (r) or append (a)

        """
        self._cpp_object = cpp.io.HDF5File(mpi_comm, filename, mode)

    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        return self._cpp_object.close()

    def get_mpi_atomicity(self) -> bool:
        """Get atomicity of the HDF5 file"""
        return self._cpp_object.get_mpi_atomicity()

    def set_mpi_atomicity(self, atomicity: bool) -> None:
        """Set atomicity of the HDF5 file"""
        self._cpp_object.set_mpi_atomicity(atomicity)

    def close(self) -> None:
        """Close file"""
        self._cpp_object.close()

    def write(self, o, name, t=None) -> None:
        """Write object to file"""
        o_cpp = getattr(o, "_cpp_object", o)
        if t is None:
            self._cpp_object.write(o_cpp, name)
        else:
            self._cpp_object.write(o_cpp, name, t)

    def read_mvc(self, mesh, name: str = ""):
        # FIXME: figure type out from file (or pass string)  and return?
        raise NotImplementedError("General MVC read function not implemented.")

    # def read_mvc(self, mesh, type: str, name: str = ""):
    #     # FIXME: return appropriate MVC based on type string
    #     raise NotImplementedError("General MVC read function not implemented.")

    # ----------------------------------------------------------

    # FIXME: implement a common function for multiple types

    def read_mvc_size_t(self, mesh, name: str = ""):
        """Read MeshValueCollection of type size_t"""
        return self._cpp_object.read_mvc_size_t(mesh, name)

    def read_mf_double(self, mesh, name: str = ""):
        """Read MeshFunction of type float"""
        return self._cpp_object.read_mf_double(mesh, name)

    # ----------------------------------------------------------

    def read_vector(self, mpi_comm, data_path: str,
                    use_partition_from_file: bool):
        """Read Vector"""
        return self._cpp_object.read_vector(mpi_comm, data_path,
                                            use_partition_from_file)

    def read_mesh(self, data_path: str,
                  use_partition_from_file: bool, ghost_mode):
        mesh = self._cpp_object.read_mesh(data_path, use_partition_from_file,
                                          ghost_mode)
        mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
        return mesh

    def read_function(self, V, name: str):
        """Read finite element Function from file

        Parameters
        ----------
        V
            Function space of saved function.
        name
            Name of function as saved into HDF file.
        Returns
        -------
        dolfin.function.function.Function
            Function read from file

        Note
        ----
        Parameter `V: Function space` must be the same as saved function space
        except for ordering of mesh entities.
        """

        V_cpp = getattr(V, "_cpp_object", V)
        u_cpp = self._cpp_object.read(V_cpp, name)
        return function.Function(V, u_cpp.vector)


class VTKFile:
    """Interface to VTK files
    VTK supports arbitrary order Lagrangian finite elements for the
    geometry description. XDMF is the preferred format for geometry
    order <= 2.

    """

    def __init__(self, filename: str):
        """Open VTK file
        Parameters
        ----------
        filename
            Name of the file
        """
        self._cpp_object = cpp.io.VTKFile(filename)

    def write(self, o) -> None:
        """Write object to file"""
        o_cpp = getattr(o, "_cpp_object", o)
        self._cpp_object.write(o_cpp)


class XDMFFile:
    """Interface to XDMF files
    This format is preferred on lower order geometries and for
    DG and RT function spaces.
    XDMF also allows for checkpointing of solutions and has parallel support.

    """

    # Import encoding (find better way?)
    Encoding = cpp.io.XDMFFile.Encoding

    def __init__(self, mpi_comm, filename: str, encoding=Encoding.HDF5):
        """Open XDMF file

        Parameters
        ----------
        mpi_comm
            The MPI communicator
        filename
            Name of the file
        encoding
            Encoding used for 'heavy' data when writing/appending

        """
        self._cpp_object = cpp.io.XDMFFile(mpi_comm, filename, encoding)

    def __enter__(self):
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        return self._cpp_object.close()

    def close(self) -> None:
        """Close file"""
        self._cpp_object.close()

    def write(self, o, t=None) -> None:
        """Write object to file

        Parameters
        ----------
        o
            The object to write to file
        t
            The time stamp

        """
        o_cpp = getattr(o, "_cpp_object", o)
        if t is None:
            self._cpp_object.write(o_cpp)
        else:
            self._cpp_object.write(o_cpp, t)

    # ----------------------------------------------------------

    # FIXME: implement a common function for multiple types

    # def read_mvc_size_t(self, mesh, name: str = ""):
    #     """Read MeshValueCollection of type size_t"""
    #     return self._cpp_object.read_mvc_size_t(mesh, name)

    def read_mvc_bool(self, mesh, name: str = ""):
        """Read MeshValueCollection of type bool"""
        return self._cpp_object.read_mvc_bool(mesh, name)

    def read_mvc_int(self, mesh, name: str = ""):
        """Read MeshValueCollection of type int"""
        return self._cpp_object.read_mvc_int(mesh, name)

    def read_mvc_size_t(self, mesh, name: str = ""):
        """Read MeshValueCollection of type size_t"""
        return self._cpp_object.read_mvc_size_t(mesh, name)

    def read_mvc_double(self, mesh, name: str = ""):
        """Read MeshValueCollection of type float"""
        return self._cpp_object.read_mvc_double(mesh, name)

    def read_mf_int(self, mesh, name: str = ""):
        """Read MeshFunction of type int"""
        return self._cpp_object.read_mf_int(mesh, name)

    def read_mf_size_t(self, mesh, name: str = ""):
        """Read MeshFunction of type size_t"""
        return self._cpp_object.read_mf_size_t(mesh, name)

    def read_mf_double(self, mesh, name: str = ""):
        """Read MeshFunction of type double"""
        return self._cpp_object.read_mf_double(mesh, name)

    # ----------------------------------------------------------

    def read_mesh(self, ghost_mode):
        mesh = self._cpp_object.read_mesh(ghost_mode)
        mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
        return mesh

    def read_mesh_data(self, mpi_comm) -> typing.Tuple[cpp.mesh.CellType, numpy.ndarray,
                                                       numpy.ndarray, typing.List[int]]:
        """Read in mesh data

        Parameters
        ----------
        mpi_comm:
            MPI communicator
        Returns
        -------
        cell_type
            Cell type
        points
            Geometric points on each process
        cells
            Topological cells with global vertex indexing
        global_cell_indices
            List of global cell indices
        """
        return self._cpp_object.read_mesh_data(mpi_comm)

    def read_checkpoint(self, V, name: str,
                        counter: int = -1) -> function.Function:
        """Read finite element Function from checkpointing format

        Parameters
        ----------
        V
            Function space of saved function.
        name
            Name of function as saved into XDMF file.
        counter : optional
            Position of function in the file within functions of the same
            name. Counter is used to read function saved as time series.
            To get last saved function use counter=-1, or counter=-2 for
            one before last, etc.
        Note
        ----
        Parameter `V: Function space` must be the same as saved function space
        except for ordering of mesh entities.
        Returns
        -------
        dolfin.function.function.Function
            The finite element Function read from checkpoint file
        """

        V_cpp = getattr(V, "_cpp_object", V)
        u_cpp = self._cpp_object.read_checkpoint(V_cpp, name, counter)
        return function.Function(V, u_cpp.vector)

    def write_checkpoint(self, u, name: str, time_step: float = 0.0) -> None:
        """Write finite element Function in checkpointing format

        """

        o_cpp = getattr(u, "_cpp_object", u)
        self._cpp_object.write_checkpoint(o_cpp, name, time_step)
